WP_HTML_Tag_Processor: Make `get_attribute` reflect attribute set via `set_attribute`, even without updating

ockham opened the pull request:
## What?
This adds a unit test to make sure `get_attribute()` picks up an attribute added by `set_attribute()`.

## Why?
This seems like a contract we might want to enforce via a unit test.

It didn't work until recently, when https://github.com/WordPress/gutenberg/pull/46598 was merged.

## How?
See code.

We might want to consider adding an assertion like that to other `set_attribute()` tests as well; or adding a separate test case just for the `set_attribut`/`get_attribute` pair.

## Testing Instructions
See unit test.

ockham committed:
WP_HTML_Tag_Processor: Ensure that reads after writes reflect most-updated values. (#46680)

Currently there exists a data race in the tag processor when reading the attribute value
for an attribute that has previously been updated. This is due to the fact that when
reading attribute values we examine the input HTML document and overlook potential
enqueued updates. If we call `get_updated_html()` before reading then those updates are
all flushed out and our values will match what we expect, but if we skip this step then
we read stale information read during the first parse of the updated tag.

In this patch we're fixing that bug by inspecting the set of enqueued changes when
reading attribute values. If an update is already enqueued for a given attribute then
we read the value from that update instead of from the original HTML. This ensures that
the value we report is always the most-recent or most-updated value.

Co-Authored-By: Adam Zielinski <adam@adamziel.com>
Co-Authored-By: Dennis Snell <dennis.snell@automattic.com>

ockham commented:
Ah, it looks like #46598 is going to fix this.

_Maybe we can add some `get_attribute` test coverage there._

ockham commented:
Closing this one; discussing adding some test coverage over there: https://github.com/WordPress/gutenberg/pull/46598#issuecomment-1359681995

anton-vlasenko commented:
> Ah, it looks like #46598 is going to fix this.
> 
> _Maybe we can add some `get_attribute` test coverage there._

@ockham 
https://github.com/WordPress/gutenberg/pull/46598 is blocking https://github.com/WordPress/gutenberg/pull/46510.
I'd appreciate it if these tests could be added in the scope of another PR.
Thanks.

ockham commented:
> > Ah, it looks like #46598 is going to fix this.
> > _Maybe we can add some `get_attribute` test coverage there._
> 
> @ockham #46598 is blocking #46510. I'd appreciate it if these tests could be added in the scope another PR. Thanks.

Okay! I guess I can re-open this PR to add the assertion.

anton-vlasenko commented:
Thanks!
I appreciate it, @ockham.

adamziel reviewed (approved):
Good spot!

adamziel commented:
E2E failures seem irrelevant so I restarted the tests

dmsnell commented:
wanted to reach out and chat more synchronously on this but I'll leave a note here for everyone's sake.

it doesn't seem right to me to say that you are "doing it wrong" here but I've been tossing around the question of whether this is intended behavior for the tag processor. once we flush out changes by calling `get_updated_html()` I think the internal pointer advances and we're ready for the next tag.

does it reset? should it reset? I don't know.

what's the use-case for reading an attribute we just set? we had to have access to the value to set it before calling `get_updated_html()` so I'm curious why we would then turn around and ask for it back.

I think this is a behavior we didn't codify, what all happens when calling `get_updated_html()` - I guess the pointer should still point at the tag we were at because we should be able to call that function any time, but on the same note, the tag processor would need to back up and re-parse the current tag if we do go back like this.

ockham commented:
Thank you for weighing in @dmsnell! I had a hunch that this had to do with the internal pointer advancing 😊 

IMO, if there's a `get_xyy` / `set_xyz` pair, there's going to be a strong expectation that setting something _will_ result in the corresponding getter retrieving that value for the thing that was just set. Realizing that that's _not_ the case, and the reasons for it, requires deeper understanding of the inner workings of the class.

And while your reasons make sense, I think it's another example of how people's habits and expectations could be so strong that they might go as far as to even change the behavior of the class to match those expectations. This is why I think we should match them already 😬 

This is assuming that people will actually encounter a need for `set_attribute()` followed by `get_attribute()`, and admittedly, my use case was prompted by a unit test ([running the `wp-bind` directive processor on a `wp-bind:src` attribute, and verifying that it adds a `src` attribute](https://github.com/WordPress/block-hydration-experiments/pull/118/files#diff-5abc2526ed26912158f3b5e124de186cac707c6e0ec4597b115ea19c0a7cc749R25-R31)).

But here's a more real-world example: For the Interactivity API, whenever we encounter a tag, we watch out for `wp-` (attribute) directives on that tag, and run the corresponding directive processors for each of them. It's absolutely possible that a tag has multiple such directives; there might be a `wp-bind:some-attribute` followed by a `wp-show` or so. They should all have access to the current tag (including adding or modifying attributes), without the need to rewind the tag pointer after they've run.

adamziel commented:
> once we flush out changes by calling get_updated_html() I think the internal pointer advances and we're ready for the next tag.

@dmsnell  `get_updated_html()` explicitly rewinds the internal pointer after flushing the changes specifically to stay at the same tag. I built this behavior with `__toString()` in mind as an innocent `$p.''` should not consume the next tag. Since `get_updated_html()` is just a human-readable name for `__toString()`, I think it's safe to say that not advancing the pointer to the next tag is a defined and expected behavior.

dmsnell commented:
so @adamziel are you saying we have a bug then?
maybe I misunderstood the whole thing - I initially thought these tests were failing and @ockham was giving us a failing test, but maybe I was confused when I first looked.

if that's the case this is good to go I think.

dmsnell commented:
okay well I _thought_ there was bug here, because if you slightly change the order of operations @ockham you will discover this still isn't quite what you expect.

```php
$p = new WP_HTML_Tag_Processor( '<div>' );
$p->next_tag();
$p->set_attribute( 'id', 'test' );
$p->get_attribute( 'id' ) === NULL;
```

> it's another example of how people's habits and expectations could be so strong that they might go as far as to even change the behavior of the class to match those expectations.

In the docblock comments we have language about attribute updates being enqueued. this makes me wonder if we should rename these functions `$p->enqueue_set_attribute`

Or just deal with it.

Or add a call to `get_updated_html()` inside `get_attribute()` to flush out pending changes, which is also something that would be nice to avoid.

🤔

so we either change the class to match those expectations or make it more obvious that those expectations shouldn't apply. this class no more serves HTTP requests than it provides a DOM API.

there's a problem with immediately updating the document once we call `set_attribute` and that's that we may still have more `set_attribute` calls which might change the same attribute or make changes to a location before the one we're currently setting.

---

I think we should be able to check first if an attribute update is enqueued and return its value before looking at the parsed values which should solve this. let me prep a patch.

ockham commented:
Just to cycle back to [this comment](https://github.com/WordPress/gutenberg/pull/46680#issuecomment-1363119752):

> maybe I misunderstood the whole thing - I initially thought these tests were failing and @ockham was giving us a failing test, but maybe I was confused when I first looked.

To be clear, this is what I did initially. However, shortly after I had filed this PR, https://github.com/WordPress/gutenberg/pull/46598 was merged, and after I rebased the PR, the original error was gone -- https://github.com/WordPress/gutenberg/pull/46598 had fixed it 😊

ockham commented:
> I think we should be able to check first if an attribute update is enqueued and return its value before looking at the parsed values which should solve this.

I think this sounds good!

> let me prep a patch.

I saw that you've pushed a commit to add the failing test (9e38e42), but I guess you haven't had time yet to work on the fix. I can do this in the next couple of days unless you'll beat me to it 😊

ockham commented:
> > let me prep a patch.
> 
> I saw that you've pushed a commit to add the failing test ([9e38e42](https://github.com/WordPress/gutenberg/commit/9e38e4245b85ebfaa5f50404cc0085f4bfbe52a7)), but I guess you haven't had time yet to work on the fix. I can do this in the next couple of days unless you'll beat me to it 😊

Fixed in b559cd4. Still need to add logic (and test coverage) for `class` attributes.

ockham commented on lib/experimental/html/class-wp-html-tag-processor.php:
```diff
@@ -1397,6 +1397,24 @@ public function get_attribute( $name ) {
 
 		$comparable = strtolower( $name );
+
+		if ( 'class' === $name ) {
+			$this->class_name_updates_to_attributes_updates();
+		}
```
For `$this->attribute_updates` (below) it makes sense (and is easy enough) to only evaluate the updates for the attribute we're interested in (if any) and to ignore updates for all other attributes. For class name updates OTOH, we would need to replicate most of the logic from `class_name_updates_to_attributes_updates` here, so we might as well just call that function and have classname updates "promoted" to attribute updates.

cc/ @adamziel since the PHPDoc for `class_name_updates_to_attributes_updates` says 

```
	 * This method is only meant to run right before the attribute updates are applied.
	 * The behavior in all other cases is undefined.
```

Although it looks okay to me to call it here?

dmsnell replied:
I think we need to call `$this->apply_attribute_updates()` here as well to ensure they get applied before any successive `class` updates occur, lest we lose enqueued class name updates.

ockham replied:
Ah yeah, you're right! While `class_name_updates_to_attributes_updates` does take an existing `class` attribute into account, it only looks at `$this->attributes['class']` to do so -- but not into `$lexical_updates`: https://github.com/WordPress/gutenberg/blob/777eaf2a34d11d3cce8456db85151141cc95a071/lib/experimental/html/class-wp-html-tag-processor.php#L1134-L1136

So we do need to call `$this->apply_attribute_updates()` as you say (unless we wanna change `class_name_updates_to_attributes_updates` to take `$lexical_updates['class']` into account -- but I guess we'd rather not).

I'll write a test case and will add a `$this->apply_attribute_updates()` call 👍

ockham replied:
Note that this will change existing semantics of `add_class` and `set_attribute` -- see: https://github.com/WordPress/gutenberg/blob/777eaf2a34d11d3cce8456db85151141cc95a071/phpunit/html/wp-html-tag-processor-test.php#L796-L828

Adding the `apply_attribute_updates` call thus breaks that test (plus a few other ones, going to look into those now).

ockham replied:
<details>
<summary>
Considering something like this
</summary>

```diff
diff --git a/lib/experimental/html/class-wp-html-tag-processor.php b/lib/experimental/html/class-wp-html-tag-processor.php
index e9c42bf419..9edd6a4105 100644
--- a/lib/experimental/html/class-wp-html-tag-processor.php
+++ b/lib/experimental/html/class-wp-html-tag-processor.php
@@ -1126,14 +1126,24 @@ class WP_HTML_Tag_Processor {
         * @see $lexical_updates
         */
        private function class_name_updates_to_attributes_updates() {
-               if ( count( $this->classname_updates ) === 0 || isset( $this->lexical_updates['class'] ) ) {
+               if ( count( $this->classname_updates ) === 0 ) {
                        $this->classname_updates = array();
                        return;
                }
 
-               $existing_class = isset( $this->attributes['class'] )
-                       ? substr( $this->html, $this->attributes['class']->value_starts_at, $this->attributes['class']->value_length )
-                       : '';
+               if ( isset( $this->lexical_updates['class'] ) ) {
+                       $existing_class_attr  = trim( $this->lexical_updates['class']->text );
+                       $existing_class_value = substr( $existing_class_attr, strlen( 'class' ) + 2, -1 );
+                       $existing_class       = html_entity_decode( $existing_class_value );
+               } elseif ( isset( $this->attributes['class'] ) ) {
+                       $existing_class = substr(
+                               $this->html,
+                               $this->attributes['class']->value_starts_at,
+                               $this->attributes['class']->value_length
+                       );
+               } else {
+                       $existing_class = '';
+               }
 
                /**
                 * Updated "class" attribute value.
```
</details>

But that will also change the semantics (two test cases -- both of which are about the `add_class`/`set_attribute` interaction AFAICS).

ockham replied:
> Adding the `apply_attribute_updates` call thus breaks that test (plus a few other ones, going to look into those now).

So the other tests affected are (the also fairly recently added) [`test_get_attribute_reflects_added_class_names_before_they_are_updated`](https://github.com/WordPress/gutenberg/pull/46680/files#diff-ea212be64658e9e6c7c4a6ac779899518ba7c773f4b4aeb8e022889274f0b6d4R551-R557), [`test_get_attribute_reflects_added_class_names_before_they_are_updated_and_retains_classes_from_previous_add_class_calls`](https://github.com/WordPress/gutenberg/pull/46680/files#diff-ea212be64658e9e6c7c4a6ac779899518ba7c773f4b4aeb8e022889274f0b6d4R566-R574) , and [`test_get_attribute_reflects_removed_class_names_before_they_are_updated`](https://github.com/WordPress/gutenberg/pull/46680/files#diff-ea212be64658e9e6c7c4a6ac779899518ba7c773f4b4aeb8e022889274f0b6d4R615-R621).

All of them call `get_attribute( 'class' )` right after setting (or removing) a class. Turns out that the problem is that if we add `$this->apply_attributes_updates()`, this flushes updates, so `get_attribute` doesn't pick up the newly added class names from `$this->lexical_attributes`; but it also doesn't parse the newly added attributes into `$this->attributes`, so `get_attribute` can't find them there either 😬 

Getting a bit into the weeds of all that update management I guess. I'm still leaning towards [this other solution](https://github.com/WordPress/gutenberg/pull/46680/files#r1072477489), which I believe has a smaller overall footprint.

ockham replied:
I've now pushed https://github.com/WordPress/gutenberg/pull/46680/commits/1517e88e0de6401d6005a9d578a2b47b1b8a2ec4 with the changes from https://github.com/WordPress/gutenberg/pull/46680/files#r1072477489. As discussed in https://github.com/WordPress/gutenberg/pull/46680/files#r1072618982, this changed some semantics of subsequent calls of `set_attribute( 'class' )` and `add_class()`. I've updated tests accordingly in https://github.com/WordPress/gutenberg/pull/46680/commits/852b5c8472f488439af761061d0a6d5c2c434422.

adamziel replied:
I like how you managed to avoid applying the attributes updates here 👍

dmsnell replied:
do we have a reason to change the behavior? I think it's probably important to ensure that `set_attribute` and `remove_attribute` updates wipe out class builder method changes, as I can reason about what it means to call `add_class` _after_ setting the `class` attribute, but I can't reason an obvious outcome of calling `add_class` _before_ the transactional swap of `set_attribute`

adamziel replied:
> an obvious outcome of calling add_class before the transactional swap of set_attribute

It is confusing 🤔 I'd like `set_attribute` to cancel any `add_class`/`remove_class` calls that happened before. I'd be fine with an undefined behavior accompanied by a warning, too, though.

dmsnell replied:
@adamziel in my latest commit I restored that defined behavior. in other words, I'm currently thinking the change in behavior wasn't necessary and so I've removed it.

ockham replied:
Thank you for fixing the behavior @dmsnell!

FWIW, you didn't exactly _revert_ my change 😬: The _original_ behavior -- prior to my changes -- was that `set_attribute( 'class', 'abc' )` would "prevail" over `add_class( 'xyz' )` _even if the `add_class` was called **after** the `set_attribute`_, thus resulting in `class="abc"`. This was covered by this unit test (this is `trunk`!): https://github.com/WordPress/gutenberg/blob/7880f08f57ac9275c21129d1f7e1f56af40bea87/phpunit/html/wp-html-tag-processor-test.php#L796-L828

I came across this when I was addressing the (apparently) missing `$this->apply_attribute_updates()` call that [you'd pointed out](https://github.com/WordPress/gutenberg/pull/46680/files#r1071709839). I realized that simply adding that would introduce [other problems](https://github.com/WordPress/gutenberg/pull/46680/files#discussion_r1072618982), so I [worked around those](https://github.com/WordPress/gutenberg/pull/46680/files#r1072477489), noting that that changed the outcome of subsequent `set_attribute( 'class', 'abc' )` and `add_class( 'xyz' )` calls. I did like that my change would have them result in `class="abc xyz"`, if the `add_class` was called after the `set_attribute` -- I still consider this an improvement over the original behavior 👍 

OTOH, it meant that if `add_class( 'xyz' )` was called _before_ `set_attribute( 'class', 'abc' )`, [it would _also_ result in `class="abc xyz"`](https://github.com/WordPress/gutenberg/pull/46680/commits/3dacb0ed4f0cff79994a91f8f11c3db8ff8ab8ff#diff-ea212be64658e9e6c7c4a6ac779899518ba7c773f4b4aeb8e022889274f0b6d4R1046-R1073). I didn't pay enough attention to that consequence, which is clearly not the behavior a user would expect 👎 

Thanks to your change, we now have [both cases behave as expected](https://github.com/WordPress/gutenberg/pull/46680/files#diff-ea212be64658e9e6c7c4a6ac779899518ba7c773f4b4aeb8e022889274f0b6d4R1063-R1068) 🎉

dmsnell replied:
Thanks for clarifying @ockham - you were right, I didn't understand that this was broken before the PR. The intention was right, the implementation was wrong.

ockham commented:
I added `get_attribute( 'class' )` assertions to a ton of existing test cases -- not sure we want to keep them all.

If we do, we might want to add `get_attribute` assertions to existing test cases that cover non-class attribute setting.

dmsnell commented on phpunit/html/wp-html-tag-processor-test.php:
```diff
@@ -721,6 +791,11 @@ public function test_set_attribute_takes_priority_over_add_class() {
+		$this->assertSame(
+			'set_attribute',
+			$p->get_attribute( 'class' ),
+			'Calling get_attribute after updating first tag\'s attributes did not return the expected class name'
```
double quotes here? also below…

ockham replied:
1911590ae493f05ec94da96cc945152690cb2428

dmsnell commented on lib/experimental/html/class-wp-html-tag-processor.php:
```diff
@@ -1397,6 +1397,24 @@ public function get_attribute( $name ) {
+		if ( isset( $this->attribute_updates[ $comparable ] ) ) {
+			$attribute = trim( $this->attribute_updates[ $comparable ]->text );
+			// Is this a boolean attribute?
+			if ( $attribute === $comparable ) {
```
this condition isn't obvious to me. in fact I'm struggling to understand how it's working. I know we have this ambiguity where the updates are purely lexical and we've lost the semantics of the type of update by the point we're looking here, and I'm wondering about that.

setting a value to `false` is indistinguishable from removing the same attribute.
removing all remaining classes is the same too as removing the `class` attribute.
setting a value to `true` should produce an empty `text` value I think.

can we make the check more direct? or was the point of the comment to indicate that this is questionable and elicit help in clarifying?

ockham replied:
A `$attribute_update[ $comparable ]->text` string is of the shape:

1. `my-attr="myvalue"` _(and not just `myvalue`!)_ for non-boolean attributes, and
2. `enabled` for (true) boolean ones.

I wanted a criterion to return `true` for the latter (and to avoid attempting to extract a non-boolean value in that case). Comparing the normalized attribute name (`$comparable`) to `$attribute_update[ $comparable ]->text` (whatever we want to call that thing) seemed like a reliable way to do. (Another option I considered was the absence of an equals sign, but that didn't seem too elegant either.) Given the shape of that thing, I'm not sure if I can think of a more direct check 🤔 -- but happy to do so if you have any ideas. 

> or was the point of the comment to indicate that this is questionable and elicit help in clarifying?

Ah, not really, TBH 😅 FWIW, the [comment a few lines below](https://github.com/WordPress/gutenberg/pull/46680/files#diff-93724b2f6e5dbf16c2b3c0efea57fa442b572f75f325d1cb185cf614b692cf9fR1413) was meant to remind the reader of the shape of `$attribute_update[ $comparable ]->text` (once we've ruled out that it's a boolean one):

```
// $attribute is a 'name="value"' string, so we extract the value.
```

dmsnell replied:
I'm worried that this condition is going to collapse as we perform HTML content changes.  we probably need to figure out a much better way of communicating the types of updates before we start stitching and cutting HTML… 🤔

ockham replied:
Can you maybe give an example of what type of scenario you're concerned about? I can't seem to think of one off the top of my head (which might be due to silent assumptions on my part).

Are we going to allow cutting and stitching _within_ an opening tag? I thought if we limited the interface to methods like `[g|s]et_content_within_balanced_tags`, attribute changes could maybe still use the current format, since they wouldn't be affected -- unless they are for a tag within those balanced tags, but in that case, they'd need to be flushed/wiped anyway, no?

Also, is this something you'd like us to sort out already, or something that can wait until we implement that kind of modifications?

dmsnell replied:
at this point the name is bad. it's `lexical_updates` and not `attribute_updates`, so the immediate work you're wanting to do, `set_content_within_balanced_tags` will produce `->text` that doesn't match this pattern.

dmsnell replied:
as long as `$comparable` is non-empty then it may be fine as that implies it's an attribute update, vs. `$this->attribute_updates[] = new WP_HTML_Text_Replacement(…)` for content changes

new thoughts:

 - we go with this for now, but note its fragility
 - we add a semantic value to the text replacement which we can query. given the limited and transient nature of these instances I'm not worried about memory use in them.
 - review the whole write-read-write-read cycle as it feels like it's starting to get out of hand. are we avoiding `get_udpated_html()` at the cost of greatly increased complexity in the code.

@adamziel at one point created a `flush()` method (or something similar) which amounted to "flush all enqueued updates" and which we could call whenever. this would make it easier to start all query, update, and seek calls and then never worry about differential accounting as we're doing in this PR

ockham replied:
Thank you Dennis! I've started [work on `set_content_inside_balanced_tags`](https://github.com/WordPress/gutenberg/pull/47036), and I'm now understanding your thoughts a lot better 😊 

I'm thinking to play a bit more with https://github.com/WordPress/gutenberg/pull/47036 to see if I can massage the updating logic into a certain shape that will be compatible with what I'm trying to do here. I _think_ we have a realistic shot at this; see my notes in that PR's description 😊 If that works out and I can limit the changes transparently to `private` methods, I might try to get those in separately, in order to unblock this PR.

adamziel replied:
```suggestion
			/*
			 * When the lexical update contains just the attribute name,
			 * the value becomes a boolean true. For example:
			 * 
			 * ```php
			 * $p = new WP_HTML_Tag_Processor('<input type="checkbox" />');
			 * $p->set_attribute('checked', true); 
			 * // The lexical update contains just the string "checked"
			 * 
			 * $p->get_attribute('checked');
			 * // returns true thanks to the condition below
			 * 
			 * echo $p->get_updated_html();
			 * // <input type="checkbox" checked />
			 * ```
			 */
			if ( $attribute === $comparable ) {
```

dmsnell replied:
@adamziel what did you mean to communicate by including this snippet?

dmsnell reviewed (commented):
just got back from AFK and am pondering this again. thanks for updating the patch. I think it's going to work out, but I feel like it's enough that I can't make an adequate response to your updates without sitting on it another day or two; I hope that's okay.

thanks for driving these changes.

ockham commented:
> I think it's going to work out, but I feel like it's enough that I can't make an adequate response to your updates without sitting on it another day or two; I hope that's okay.

Absolutely! I'm not in a rush (and will be AFK tomorrow anyway) 😊

dmsnell reviewed (commented):
I've finally worked through this to try and understand and have basically came up with what you have here, except for a couple spots I think are missed.

We need to clear out any enqueued updates inside of `remove_attribute()` when calling that on an attribute that has already been updated.

We also need to ensure we don't flush out _some_ `class` updates and then leave open an opportunity to mess up `add_class` and `remove_class` calls by partially applying those updates. A call to `apply_attributes_updates()` should do it.

In my patch I've also renamed a few variables that confused me while trying to read through the algorithm. I may propose those in another PR.

<details><summary>my patch</summary>

```diff
diff --git a/lib/experimental/html/class-wp-html-tag-processor.php b/lib/experimental/html/class-wp-html-tag-processor.php
index c7bb67ba27..1ac987f598 100644
--- a/lib/experimental/html/class-wp-html-tag-processor.php
+++ b/lib/experimental/html/class-wp-html-tag-processor.php
@@ -254,7 +254,7 @@ class WP_HTML_Tag_Processor {
 	 * @since 6.2.0
 	 * @var string
 	 */
-	private $updated_html = '';
+	private $html_with_updates = '';
 
 	/**
 	 * How many bytes from the original HTML document were already read.
@@ -271,7 +271,7 @@ class WP_HTML_Tag_Processor {
 	 * @since 6.2.0
 	 * @var int
 	 */
-	private $updated_bytes = 0;
+	private $processed_bytes = 0;
 
 	/**
 	 * Byte offset in input document where current tag name starts.
@@ -1264,9 +1264,9 @@ class WP_HTML_Tag_Processor {
 		usort( $this->lexical_updates, array( self::class, 'sort_start_ascending' ) );
 
 		foreach ( $this->lexical_updates as $diff ) {
-			$this->updated_html .= substr( $this->html, $this->updated_bytes, $diff->start - $this->updated_bytes );
-			$this->updated_html .= $diff->text;
-			$this->updated_bytes = $diff->end;
+			$this->html_with_updates .= substr( $this->html, $this->processed_bytes, $diff->start - $this->processed_bytes );
+			$this->html_with_updates .= $diff->text;
+			$this->processed_bytes = $diff->end;
 		}
 
 		foreach ( $this->bookmarks as $bookmark ) {
@@ -1335,8 +1335,8 @@ class WP_HTML_Tag_Processor {
 
 		// Point this tag processor before the sought tag opener and consume it.
 		$this->parsed_bytes  = $this->bookmarks[ $bookmark_name ]->start;
-		$this->updated_bytes = $this->parsed_bytes;
-		$this->updated_html  = substr( $this->html, 0, $this->updated_bytes );
+		$this->processed_bytes = $this->parsed_bytes;
+		$this->html_with_updates  = substr( $this->html, 0, $this->processed_bytes );
 		return $this->next_tag();
 	}
 
@@ -1397,6 +1397,26 @@ class WP_HTML_Tag_Processor {
 		}
 
 		$comparable = strtolower( $name );
+
+		// Check if we've queued an update, and if so, read that one first
+		if ( array_key_exists( $comparable, $this->lexical_updates ) ) {
+			$update = $this->lexical_updates[ $comparable ]->text;
+
+			if ( '' === $update ) {
+				return false;
+			}
+
+			$equals_at = strpos( $update, '=' );
+			if ( false === $equals_at ) {
+				return true;
+			}
+
+			// Because we serialize all attributes with double-quotes, we can assume the syntax here.
+			$updated_value = substr( $update, $equals_at + 2, strlen( $update ) - $equals_at - 3 );
+
+			return html_entity_decode( $updated_value );
+		}
+
 		if ( ! isset( $this->attributes[ $comparable ] ) ) {
 			return null;
 		}
@@ -1647,6 +1667,11 @@ class WP_HTML_Tag_Processor {
 		 * @see https://html.spec.whatwg.org/multipage/syntax.html#attributes-2:ascii-case-insensitive
 		 */
 		$name = strtolower( $name );
+
+		if ( isset( $this->lexical_updates[ $name ] ) ) {
+			unset( $this->lexical_updates[ $name ] );
+		}
+
 		if ( $this->is_closing_tag || ! isset( $this->attributes[ $name ] ) ) {
 			return false;
 		}
@@ -1723,37 +1748,45 @@ class WP_HTML_Tag_Processor {
 	 * @return string The processed HTML.
 	 */
 	public function get_updated_html() {
-		// Short-circuit if there are no new updates to apply.
-		if ( ! count( $this->classname_updates ) && ! count( $this->lexical_updates ) ) {
-			return $this->updated_html . substr( $this->html, $this->updated_bytes );
+		$requires_no_updating = 0 === count( $this->classname_updates ) && 0 === count( $this->lexical_updates );
+
+		// If we have no updates and haven't started updating, return the input to avoid the copy.
+		if ( $requires_no_updating && $this->processed_bytes === 0 ) {
+			return $this->html;
+		}
+
+		// Otherwise, if we have no updates then complete copying the input to the updated document and return.
+		if ( $requires_no_updating && $this->processed_bytes > 0 ) {
+			return $this->html_with_updates . substr( $this->html, $this->processed_bytes );
 		}
 
 		// Otherwise: apply the updates, rewind before the current tag, and parse it again.
-		$delta_between_updated_html_end_and_current_tag_end = substr(
+		$content_up_to_opened_tag_name = $this->html_with_updates . substr(
 			$this->html,
-			$this->updated_bytes,
-			$this->tag_name_starts_at + $this->tag_name_length - $this->updated_bytes
+			$this->processed_bytes,
+			$this->tag_name_starts_at + $this->tag_name_length - $this->processed_bytes
 		);
-		$updated_html_up_to_current_tag_name_end            = $this->updated_html . $delta_between_updated_html_end_and_current_tag_end;
 
-		// 1. Apply the attributes updates to the original HTML
+		// 1. Apply the edits by flushing them to $this->updated_html and updating the updated_bytes
 		$this->class_name_updates_to_attributes_updates();
 		$this->apply_attributes_updates();
 
-		// 2. Replace the original HTML with the updated HTML
-		$this->html          = $this->updated_html . substr( $this->html, $this->updated_bytes );
-		$this->updated_html  = $updated_html_up_to_current_tag_name_end;
-		$this->updated_bytes = strlen( $this->updated_html );
-
-		// 3. Point this tag processor at the original tag opener and consume it
+		// 2. Replace the original HTML with the updated HTML so that it maintains consistency with our updates.
+		$this->html              = $this->html_with_updates . substr( $this->html, $this->processed_bytes );
+		$this->html_with_updates = $content_up_to_opened_tag_name;
+		$this->processed_bytes   = strlen( $this->html_with_updates );
 
-		/*
+		/* 3. Point this tag processor at the original tag opener and consume it
+		 *
+		 * By going back to the start of the current open tag we stay "in the same location" in
+		 * the document and can continue to modify the attributes of the current tag.
+		 *
 		 * When we get here we're at the end of the tag name, and we want to rewind to before it
 		 * <p>Previous HTML<em>More HTML</em></p>
 		 *                 ^  | back up by the length of the tag name plus the opening <
 		 *                 \<-/ back up by strlen("em") + 1 ==> 3
 		 */
-		$this->parsed_bytes = strlen( $updated_html_up_to_current_tag_name_end ) - $this->tag_name_length - 1;
+		$this->parsed_bytes = strlen( $content_up_to_opened_tag_name ) - $this->tag_name_length - 1;
 		$this->next_tag();
 
 		return $this->html;
diff --git a/phpunit/html/wp-html-tag-processor-test.php b/phpunit/html/wp-html-tag-processor-test.php
index 6750410f26..36b6e0ca8a 100644
--- a/phpunit/html/wp-html-tag-processor-test.php
+++ b/phpunit/html/wp-html-tag-processor-test.php
@@ -8,6 +8,12 @@
 
 require_once __DIR__ . '/../../lib/experimental/html/index.php';
 
+if ( ! class_exists( 'WP_UnitTestCase' ) ) {
+	class WP_UnitTestCase extends \PHPUnit\Framework\TestCase {}
+
+	function esc_attr( $s ) { return str_replace( ['<', '"'], ['&lt;', '&quot;'], $s ); }
+}
+
 /**
  * @group html
  *
@@ -229,6 +235,24 @@ class WP_HTML_Tag_Processor_Test extends WP_UnitTestCase {
 		$this->assertNull( $p->get_attribute_names_with_prefix( 'data-' ), 'Accessing attributes of a non-existing tag did not return null' );
 	}
 
+	public function test_reads_updates_before_they_are_applied() {
+		$p = new WP_HTML_Tag_Processor( '<div data-foo="bar">Test</div>' );
+		$p->next_tag();
+		$this->assertNull( $p->get_attribute( 'test-id' ) );
+
+		$p->set_attribute( 'test-id', true );
+		$this->assertTrue( $p->get_attribute( 'test-id' ) );
+
+		$p->set_attribute( 'test-id', false );
+		$this->assertNull( $p->get_attribute( 'test-id' ) );
+
+		$p->set_attribute( 'test-id', '5 < 8' );
+		$this->assertEquals( '5 < 8', $p->get_attribute( 'test-id' ) );
+
+		$p->get_updated_html();
+		$this->assertEquals( '5 < 8', $p->get_attribute( 'test-id' ) );
+	}
+
 	/**
 	 * @ticket 56299
 	 *
```
</details>

one thing I realized is that we need to be careful when we apply non-attribute updates because `get_updated_html()` and `apply_attributes_updates()` assume that any lexical diffs occur _after_ the current pointer of processed bytes (`$this->updated_html`). one of the dangers I suspected is real, that if we apply a modification across bookmarks we need to make sure we flush out any pending attribute updates and then that we properly account for the fact that those updates cross the pointer boundaries.

ockham commented:
> I've finally worked through this to try and understand and have basically came up with what you have here, except for a couple spots I think are missed.

🎉 

> We need to clear out any enqueued updates inside of `remove_attribute()` when calling that on an attribute that has already been updated.

Ah, good point! I've added test coverage for this scenario in https://github.com/WordPress/gutenberg/pull/46680/commits/7f003ae516f43fd1910e8ede83e068c10d274f34 and https://github.com/WordPress/gutenberg/pull/46680/commits/466bf84d537a9d46b68f1e7708f3058a7ed28815, and fixed the implementation to make those pass in https://github.com/WordPress/gutenberg/pull/46680/commits/8754cf7c987d188b32de2f1a81e116c1e392523e and https://github.com/WordPress/gutenberg/pull/46680/commits/825b53f91c16a3690fe6aa90f10c4d7613529979, respectively.

> We also need to ensure we don't flush out _some_ `class` updates and then leave open an opportunity to mess up `add_class` and `remove_class` calls by partially applying those updates. A call to `apply_attributes_updates()` should do it.

I'll have a look at that next; I _thought_ that I'd considered that when first writing the PR but decided that it wasn't needed -- but I'll double-check now 😊 

> In my patch I've also renamed a few variables that confused me while trying to read through the algorithm. I may propose those in another PR.
> my patch
> [...]

I like those! I got a bit confused myself by names like `$updated_bytes` or `$delta_between_updated_html_end_and_current_tag_end`. But yeah, might be worth a separate PR, so we're able to isolate changes better.
 
> one thing I realized is that we need to be careful when we apply non-attribute updates because `get_updated_html()` and `apply_attributes_updates()` assume that any lexical diffs occur _after_ the current pointer of processed bytes (`$this->updated_html`). one of the dangers I suspected is real, that if we apply a modification across bookmarks we need to make sure we flush out any pending attribute updates and then that we properly account for the fact that those updates cross the pointer boundaries.

Yeah, that sounds pretty much like [the problem that I was facing yesterday](https://github.com/WordPress/gutenberg/pull/47036#issuecomment-1384237914)... 😅

adamziel commented on lib/experimental/html/class-wp-html-tag-processor.php:
```diff
@@ -1647,7 +1681,12 @@ public function remove_attribute( $name ) {
 		$name = strtolower( $name );
-		if ( $this->is_closing_tag || ! isset( $this->attributes[ $name ] ) ) {
+
+		if ( ! isset( $this->attributes[ $name ] ) ) {
+			if ( isset( $this->lexical_updates[ $name ] ) ) {
+				unset( $this->lexical_updates[ $name ] );
+			}
+
```
I don't think it would process this case correctly?

```php
$p = new WP_HTML_Tag_Processor('<div id="a"></div>');
$p->next();
$p->set_attribute('id', 'b');
$p->remove_attribute('id');
```

How about changing the condition like this?
```suggestion
		if ( isset( $this->lexical_updates[ $name ] ) ) {
			unset( $this->lexical_updates[ $name ] );
		}
			
		if ( ! isset( $this->attributes[ $name ] ) ) {

```

This way any enqueued updates will be removed regardless of whether the attribute already existed.

adamziel replied:
Also should this method call `class_name_updates_to_attributes_updates()` now?

ockham replied:
> I don't think it would process this case correctly?

No, that should work! 😄 

<details>
<summary>
You can try this patch
</summary>

```diff
diff --git a/phpunit/html/wp-html-tag-processor-test.php b/phpunit/html/wp-html-tag-processor-test.php
index 662749a523..55e5ba8c24 100644
--- a/phpunit/html/wp-html-tag-processor-test.php
+++ b/phpunit/html/wp-html-tag-processor-test.php
@@ -605,6 +605,23 @@ class WP_HTML_Tag_Processor_Test extends WP_UnitTestCase {
                $this->assertSame( self::HTML_SIMPLE, $p->get_updated_html() );
        }
 
+       /**
+        * @ticket 56299
+        *
+        * @covers set_attribute
+        * @covers remove_attribute
+        * @covers get_attribute
+        * @covers get_updated_html
+        */
+       public function test_get_attribute_reflects_set_attribute_followed_by_remove_attribute_for_existing_attribute_before_it_is_updated() {
+               $p = new WP_HTML_Tag_Processor( self::HTML_SIMPLE );
+               $p->next_tag();
+               $p->set_attribute( 'id', 'test-value' );
+               $p->remove_attribute( 'id' );
+               $this->assertNull( $p->get_attribute( 'id' ) );
+               $this->assertSame( '<div ><span id="second">Text</span></div>', $p->get_updated_html() );
+       }
+
        /**
         * @ticket 56299
         *
```

</details>

(I'll push a commit with that test.)

The idea is as follows:
- If the attribute is not set in the HTML, we remove any update that would add it.
- If the attribute _is_ set in the HTML, we override any updates with a new one that will remove it, when applied. That latter part is handled by existing code a couple lines further down (that's not touched here): https://github.com/WordPress/gutenberg/blob/340b51d44558bf97402c46150887d88ff796404c/lib/experimental/html/class-wp-html-tag-processor.php#L1654-L1669

ockham replied:
> Also should this method call `class_name_updates_to_attributes_updates()` now?

Ah, good point! I think it should indeed 🤔 I'll also add some test coverage !

ockham replied:
Test coverage in d739f31d14.

I think we can actually get away without `class_name_updates_to_attributes_updates` -- clearing `this$->classname_updates` should be enough: 06245b3726

adamziel commented on lib/experimental/html/class-wp-html-tag-processor.php:
```diff
@@ -1116,24 +1116,31 @@ private function after_tag() {
-			: '';
+		if ( isset( $this->lexical_updates['class'] ) ) {
+			$existing_class_attr  = trim( $this->lexical_updates['class']->text );
+			$existing_class_value = substr( $existing_class_attr, strlen( 'class' ) + 2, -1 );
```
There's also a (silly) case of `$p->set_attribute('class', true);` where `$existing_class_attr` will only contain `class`.

ockham replied:
Test coverage in https://github.com/WordPress/gutenberg/pull/46680/commits/4c225277de5bdaddf5d95a3164fbbe523ec2b070, fix in https://github.com/WordPress/gutenberg/pull/46680/commits/c48ba893e92a44b3a43691c7905fdf4422a0ab74.

ockham commented:
Thank you for reviewing @adamziel! I believe I've addressed all you feedback now 😊

adamziel commented on phpunit/html/wp-html-tag-processor-test.php:
```diff
@@ -516,12 +516,159 @@ public function data_set_attribute_escapable_values() {
+		$p = new WP_HTML_Tag_Processor( self::HTML_SIMPLE );
+		$p->next_tag();
+		$p->set_attribute( 'test-attribute', 'test-value' );
+		$this->assertSame( 'test-value', $p->get_attribute( 'test-attribute' ) );
```
The convention in WordPress core, where these tests will eventually be merged, is to provide an error message every time a test fails. This means adjusting all the new `assert` calls as follows:

```suggestion
		$this->assertSame( 'test-value', $p->get_attribute( 'test-attribute' ), "get_attribute() did not return "test-value" for the attribute "test-attribute" after "test-value" was assigned using set_attribute()." );
```

I know 😬

adamziel reviewed (approved):
I left a note about the unit tests but it looks good to me otherwise. Thank you @ockham for addressing all the feedback!

dmsnell committed:
Revert behavioral change, refactor a few things

 - Presereed the original behavior when class-builder methods and attribute-setters
   for the `class` attribute coincide.

 - Renamed the enqueued-getter to `get_enqueued_attribute_value()` to focus
   on its semantic purpose instead of the implementation steps.

 - Un-doc-blocks some accidental docblock comments.

 - Reworded language use in a comment and removed "OTOH" as an unexplained abbreviation.

dmsnell reviewed (approved):
sorry for the last-minute changes but I pushed 5e25270e66d54dba8a8aa6e9909402523001d782 in which I rearranged a few things stylistically and also reverted the behavioral change, because I find it confusing to preserve `add_class`/`remove_class` calls _after_ `set_attribute` or `remove_attribute` has occurred.

this is negotiable, but I didn't see any clear reason why we changed the behavior (though I may have missed it while reading through the comments). if the reason is that it was a side-effect of rearranging the code then I think the change I've proposed is not worse than what it was before or after that change.

now, when calling the transactional swaps `set_attribute` or `remove_attribute` for `class`, all enqueued class-builder changes are wiped out and the new value goes in. any calls to the builder after this point augments the updated value. I believe this maps most closely to how successive edits to any other attribute work.

let me know how you feel. if you don't like what I changed then we can talk about it. the stylistic things are also up for grabs, but I think they are mostly new paint on the framework you built.

dmsnell commented:
Squashed earlier work and rebased against latest `trunk` from 5e25270e66d54dba8a8aa6e9909402523001d782 to fafafce0d5d667a4a883ccb12bebd3fb03b1e59d

Second commit on the rebase contains the changes I made in my work today, which we can discuss, revert, or accept.

adamziel commented on lib/experimental/html/class-wp-html-tag-processor.php:
```diff
@@ -1378,50 +1380,66 @@ private static function sort_start_ascending( $a, $b ) {
-
-		if ( ! isset( $this->lexical_updates[ $comparable ] ) ) {
-			return null;
+	private function get_enqueued_attribute_value( $comparable_name ) {
```
I love this change

adamziel replied:
Actually... it confuses me how `add_class` etc updates are not considered "enqueued" by this method. It may be too much, but here's an idea – what if we renamed `lexical_updates` to `enqueued_updates`? That would be a clear-cut distinction of what's enqueued and what isn't. Also, it would clarify that updates haven't been applied yet. Finally, it would draw a clear line between enqueued updates and utility buffers such as class names and parsed attributes.

ockham replied:
Renaming `lexical_updates` to `enqueued_updates` works for me. (I'd do it in a follow-up PR though, this one has grown big enough already 😬)

dmsnell replied:
I'm not bothered by the fact that class-builder updates aren't considered here. in fact there's a circularity problem if we make it work that way, because `class_updates_to_attribute_updates` needs to read this method.

as a private method it's not supposed to be exposed either. we can add a note about `class` being special.

as per `enqueued_updates` vs. `lexical_updates` I don't know. I don't see much win but I don't see a loss. the difference is that when we renamed it to `lexical_updates` it was meant to better reflect that these are textual diffs being applies without any other semantics. `enqueued_updates` somewhat hints that there might be more structure or semantic value associated with the updates when in fact there aren't and this whole system _so far_ has been built around the idea that they are intentionally dumb (well, apart from reprising them in this PR 🙃).

ockham replied:
I'm fine either way (and more than happy to defer the rename 😄)

adamziel commented on lib/experimental/html/class-wp-html-tag-processor.php:
```diff
@@ -1475,6 +1495,17 @@ public function get_attribute( $name ) {
 
+		/*
+		 * This flag distinguishes an attribute with no value
+		 * from an attribute with an empty string value. For
```
I must have missed the discussion – we could distinguish between NULL values, false values, and empty strings. Is the `is_true` flag a poka yoke/a developer experience affordance to make it more convenient to reason about? Or is there more to it?

dmsnell replied:
There's more to it and I'm pretty sure it'd come back to you if you looked, as it was the case for me.

First of all, there's no `false` attribute value in HTML. Only `true`. An attribute is either a boolean `true` or it's absent. The DOM API doesn't even expose this. There is no way to differentiate an attribute with an empty string value and a true value, but `setAttribute( 'wp-test', '' )` omits the string value when creating the HTML representation of the node. The DOM API is more matter-of-fact/doesn't matter about it than we are, but we're attempting to represent syntactical equivalence rather than semantic.

https://user-images.githubusercontent.com/5431237/214435276-745da71e-d8fb-481d-860b-3d584a89b3f7.mov

In other words this exists to differentiate cases like `enabled` and `enabled=enabled` and `enabled=""`

adamziel commented on lib/experimental/html/class-wp-html-tag-processor.php:
```diff
@@ -1700,6 +1731,14 @@ public function set_attribute( $name, $value ) {
+		 * Any calls to update the `class` attribute directly should wipe out any
+		 * enqueued class changes from `add_class` and `remove_class`.
+		 */
+		if ( 'class' === $comparable_name && ! empty( $this->classname_updates ) ) {
```
👍

adamziel commented on phpunit/html/wp-html-tag-processor-test.php:
```diff
@@ -772,13 +1035,15 @@ public function test_remove_class_preserves_whitespaces_when_there_is_a_class_at
 			'<div class="   main" id="first"><span class="not-main bold with-border" id="second">Text</span></div>',
 			$p->get_updated_html()
 		);
+		$this->assertSame( '   main', $p->get_attribute( 'class' ) );
```
@ockham there's a few more assertions without an explicit error message as a third attribute.

ockham replied:
https://github.com/WordPress/gutenberg/pull/46680/commits/0110bcf04f38e1232b82cd4be9908819b84627af

adamziel commented:
I think it's mostly there – I left a few final notes 👍  (have we been repeating this for two weeks now? :D)

ockham commented on lib/experimental/html/class-wp-html-tag-processor.php:
```diff
@@ -1378,50 +1380,66 @@ private static function sort_start_ascending( $a, $b ) {
+	 * Return the enqueued value for a given attribute, if one exists.
+	 *
+	 * Enqueued updates can take different data types:
+	 *  - If an update is enqueued and is boolean, the return will be `true`
+	 *  - If an update is otherwise enqueued, the return will be the string value of that update.
+	 *  - If an attribute is enqueued to be removed, the return will be `null` to indicate that.
+	 *  - If no updates are enqueued, the return will be `false` to differentiate from "removed."
```
The semantics of this are a bit tricky 🤔 It's probably inherent to the fact that on top of HTML attribute semantics (which are reflected by`get_attribute`), we also need a value to mean "no updates", but I'm not sure if `false` would be my preferred choice. (I was a bit confused when first reading conditionals like [this](https://github.com/WordPress/gutenberg/pull/46680/commits/fafafce0d5d667a4a883ccb12bebd3fb03b1e59d#diff-93724b2f6e5dbf16c2b3c0efea57fa442b572f75f325d1cb185cf614b692cf9fR1135).)

(Admittedly, I can't think of a much better semantics off the top of my head.)

dmsnell replied:
I left the commit with the thought of adding class constants which would reflect better. Left them as a follow-up to be sure it's a good idea.

If `null` weren't already used to mean "no attribute" I'd use that here, but in the case of updates, we can _update_ and attribute to _remove it_ which would indicate `null` 😆 

Thankfully `false` is not possible as an attribute value so I reached for it out of convenience.

The one issue with constants is that we have to be careful not to conflate them with legitimate attribute values, so strings are out, leaving us back with `null` and `false`. We could also create empty classes as a faux enum, but then we're adding potentially needless complexity and overhead.

ockham replied:
Yeah, agree that `false` seems like a natural contender, since it was available. It's still not the most intuitive thing, but since this is a private method, and you've nicely documented it, let's go with it 👍

adamziel replied:
@dmsnell constants like ‚REMOVE_ATTR = null’ or ‚NO_UPDATES = false’ would make this easier to reason about AND preserve the semantics you described

dmsnell replied:
that's one thing I thought of @adamziel but then the worry I had was how those obscure the values. unlike with `ADD_CLASS` and `REMOVE_CLASS`, we actually use the return values here and get no type safety.

for example, this code looks clear…

```php
if ( self::NO_UPDATES === $this->get_enqueued_attribute_value( $name ) ) {
	…
}
```

but then because these constants look like real expected values I think there's some obscurity in this code…

```php
if ( null === $this->get_enqueuedd_attribute_value( $name ) ) {
	…
}
```

just noting here that I had some reticence in jumping to constants because of the way they hide the values that we do end up using as those primitive values.

ockham commented on lib/experimental/html/class-wp-html-tag-processor.php:
```diff
@@ -1378,50 +1380,66 @@ private static function sort_start_ascending( $a, $b ) {
+		 *           ¹²    ³
+		 *        1. Equals is here (E).
+		 *        2. Double-quoting starts one after the equals sign (E + 1).
+		 *        3. Double-quoting ends at the last character in the update (Total Length - E - 3 syntax characters).
+		 */
+		$enqueued_value = substr( $enqueued_text, $equals_at + 2, strlen( $enqueued_text ) - $equals_at - 3 );
```
Can't we use

```suggestion
		 *        3. Double-quoting ends at the last character in the update.
		 */
		$enqueued_value = substr( $enqueued_text, $equals_at + 2, -1 );
```

to point to the end of the string, minus one character? I feel like that's easier to follow than the `strlen( $enqueued_text ) - $equals_at - 3` computation 😬 

`substr` [docs](https://www.php.net/manual/en/function.substr.php) for the `$length` arg:

> If `length` is given and is negative, then that many characters will be omitted from the end of string

dmsnell replied:
very good idea! why didn't I think of that 😄

ockham commented on lib/experimental/html/class-wp-html-tag-processor.php:
```diff
@@ -1452,21 +1470,23 @@ public function get_attribute( $name ) {
-		 * replicate most of the logic from `class_name_updates_to_attributes_updates` here,
-		 * so we might as well just call that function and have classname updates "promoted"
-		 * to attribute updates.
+		/*
+		 * For every attribute other than `class` we can perform a quick check if there's an
+		 * enqueued lexical update whose value we should prefer over what's in the input HTML.
+		 *
+		 * The `class` attribute is special though because we expose the helpers `add_class`
+		 * and `remove_class` which form a builder for the `class` attribute, so we have to
+		 * additionally check if there are any enqueued class changes. If there are, we need
+		 * to first flush them out so can report the full string value of the attribute.
```
👍

ockham committed:
More assertion error messages

ockham commented:
Okay, I've added the missing assertion error messages. We're all green -- looks like we can finally 🚢  this 🎉 

Thank you @dmsnell @adamziel for your help! Great collab'ing with you 😊
